### PR TITLE
Added -Vault parameter to RecoveryServices.Backup cmdlets

### PR DESCRIPTION
**Pull request by dragonfly91**
## Description

Added -Vault parameter to RecoveryServices.Backup cmdlets

## Checklist

- [x] I have read the [_Submitting Changes_](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#submitting-changes) section of [`CONTRIBUTING.md`](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md)
- [x] The title of the PR is clear and informative
- [x] The appropriate [change log has been updated](https://github.com/Azure/azure-powershell/blob/preview/CONTRIBUTING.md#updating-the-change-log)
- [x] The PR does not introduce [breaking changes](https://github.com/Azure/azure-powershell/blob/preview/documentation/breaking-changes/breaking-changes-definition.md)
- [x] If applicable, the changes made in the PR have proper test coverage
- [x] For public API changes to cmdlets:
    - [x] the changes have gone through a [cmdlet design review](https://github.com/Azure/azure-powershell-cmdlet-review-pr)
    - [x] the cmdlet markdown files were [generated using the `platyPS` module](https://github.com/Azure/azure-powershell/blob/preview/documentation/development-docs/help-generation.md)

### CODE REVIEW

**Comment by dragonfly91** [comment]
@maddieclayton Created this review request here: https://github.com/Azure/azure-powershell-cmdlet-review-pr/issues/79. Please check!

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Test/ScenarioTests/AzureSql/ItemTests.ps1 [hunk 1]
```diff
@@ -42,47 +43,48 @@ function Test-AzureSqlGetItems
...
+		-Force;
 }
 
 function Test-AzureSqlGetRPs
 {
-	#Set vault context
+	# 1. Get the vault
 	$vault = Get-AzureRmRecoveryServicesVault -ResourceGroupName "sqlpaasrg" -Name "sqlpaasrn";
 	
-	# 2. Set the vault context
```
Comment (maddieclayton) [inline src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Test/ScenarioTests/AzureSql/ItemTests.ps1]: Can you have one test where you set the context with Set-AzureRmRecoveryServicesVaultContext? I assume we want this to continue to work, so we should have at least one test for it.

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/RSBackupVaultCmdletBase.cs [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+            ValueFromPipeline = true)]
+        [ValidateNotNullOrEmpty]
+        public ARSVault Vault { get; set; }
+
+        /// <summary>
+        /// Get the job PS model after fetching the job object from the service given the job ID.
+        /// </summary>
+        /// <param name="jobId">ID of the job to be fetched</param>
+        /// <returns></returns>
+        public CmdletModel.JobBase GetJobObject(string jobId, ARSVault vault = null)
```
Comment (maddieclayton) [inline src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/RSBackupVaultCmdletBase.cs]: Do the corresponding "GetJobObject/HandleCreatedJob" calls in RecoveryServicesBackupCmdletBase still work with the changes to the adapter? If not, please remove them.

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/help/Wait-AzureRmRecoveryServicesBackupJob.md [hunk 3]
```diff
@@ -98,6 +98,21 @@ Accept pipeline input: False
...
+Aliases: 
+
+Required: False
+Position: Named
+Default value: None
+Accept pipeline input: True (ByValue)
+Accept wildcard characters: False
+```
+
 ### CommonParameters
```
Comment (maddieclayton) [inline src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/help/Wait-AzureRmRecoveryServicesBackupJob.md]: Please revert the changes to the three files below this.

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Commands.RecoveryServices.Backup.Providers.csproj [hunk 4]
```diff
@@ -88,6 +88,10 @@
       <Project>{b758fec1-35c1-4f93-a954-66dd33f6e0ec}</Project>
       <Name>Commands.RecoveryServices.Backup.ServiceClientAdapter</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Commands.RecoveryServices\Commands.RecoveryServices.csproj">
```
Comment (maddieclayton) [inline src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Commands.RecoveryServices.Backup.Providers.csproj]: Why do you need to add this reference?

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/Commands.RecoveryServices.Backup.Cmdlets.csproj [hunk 5]
```diff
@@ -109,6 +110,10 @@
       <Project>{b758fec1-35c1-4f93-a954-66dd33f6e0ec}</Project>
       <Name>Commands.RecoveryServices.Backup.ServiceClientAdapter</Name>
     </ProjectReference>
+    <ProjectReference Include="..\Commands.RecoveryServices\Commands.RecoveryServices.csproj">
```
Comment (maddieclayton) [inline src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/Commands.RecoveryServices.Backup.Cmdlets.csproj]: Why do you need to add this reference?

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/RSBackupVaultCmdletBase.cs (hunk 2 above)
Comment (dragonfly91) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/RSBackupVaultCmdletBase.cs]: Yes, they still work.

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Commands.RecoveryServices.Backup.Providers.csproj (hunk 4 above)
Comment (dragonfly91) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Commands.RecoveryServices.Backup.Providers.csproj]: This project contains the definition of ARSVault which is being referenced by the methods in the provider files.

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/Commands.RecoveryServices.Backup.Cmdlets.csproj (hunk 5 above)
Comment (dragonfly91) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/Commands.RecoveryServices.Backup.Cmdlets.csproj]: This project contains the definition of ARSVault which is being referenced by the methods in this project.

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Test/ScenarioTests/AzureSql/ItemTests.ps1 (hunk 1 above)
Comment (dragonfly91) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Test/ScenarioTests/AzureSql/ItemTests.ps1]: Done.

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/help/Wait-AzureRmRecoveryServicesBackupJob.md (hunk 3 above)
Comment (dragonfly91) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/help/Wait-AzureRmRecoveryServicesBackupJob.md]: Done.

**Comment by MiYanni** [comment]
@dragonfly91 You have merge conflicts that need to be resolved.

**Comment by dragonfly91** [comment]
Fixed the merge conflicts. Please continue review!

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Commands.RecoveryServices.Backup.Providers.csproj (hunk 4 above)
Comment (cormacpayne) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Commands.RecoveryServices.Backup.Providers.csproj]: @dragonfly91 until all of the RecoveryServices modules are consolidated into one, we should avoid introducing any coupling between the types -- this prevents us from getting into scenarios where users can only use certain Backup functionality when paired with RecoveryServices cmdlets. Once the modules are consolidated, we can have a common models projects that all of the projects share and can use interchangeably without worrying about having additional modules installed and run beforehand.

Comment (cormacpayne) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Commands.RecoveryServices.Backup.Providers.csproj]: @dragonfly91 as a workaround, we could have the user provide the vault resource id, which you can parse using the [`ResourceIdentifier`](https://github.com/Azure/azure-powershell/blob/preview/src/ResourceManager/Common/Commands.ResourceManager.Common/Utilities/Models/ResourceIdentifier.cs) class, to get the vault name and resource group, which appear to be the only two properties you are using from the vault object (let me know if that's not the case)

Comment (dragonfly91) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Commands.RecoveryServices.Backup.Providers.csproj]: @cormacpayne 

When we had this discussion last time regarding merge, this point about how to address module dependency was never mentioned explicitly by the ADX team, I believe. Removing dependency on the Set-AzureRmRecoveryServicesVaultContext cmdlet is one of the biggest pain points for the customers and providing the Vault object is the most comfortable way for them. @pvrk can you comment on the workaround suggested by @cormacpayne ?

Also, this PR has been open for the last 9 days. This comment, given on the 9th day, essentially calls for a complete code revamp; all the files will have to be modified. 

I understand your commitment to make things easy for the customer. But, I feel that there is no way to anticipate the PR review process. This apparent issue in our design could have been:

1. Educated in the initial talks about merge
2. Flagged in the PR Design Review
3. Flagged in the first few days of PR Code Review.

My concern: why didn't this comment come up in any of the above steps? We have now worked on throw away code and also risk missing the release. 

Hope you understand our pain point and we need you to address this.

Comment (pvrk) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Providers/Commands.RecoveryServices.Backup.Providers.csproj]: @cormacpayne : So the idea is to use vault ID as a workaround and switch it back to object when the code merger is done? Even though if the idea is to use vault ID throughout, it came at the very last minute at a very high cost (we need to modify all the files) and I concur with @dragonfly91 here. We need such inputs as early as possible.

**Comment by dragonfly91** [comment]
Hi, updated the PR based on the comments obtained. Used VaultId instead of Vault to avoid introducing new dependencies between modules. Please review / merge!

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/RSBackupVaultCmdletBase.cs [hunk 6]
```diff
@@ -17,32 +18,44 @@ public abstract class RSBackupVaultCmdletBase : RecoveryServicesBackupCmdletBase
...
+        public string VaultId { get; set; }
+
+        /// <summary>
+        /// Location of the Recovery Services Vault.
+        /// </summary>
+        [Parameter(Mandatory = false, HelpMessage = "Location of the Recovery Services Vault.",
+            ValueFromPipeline = true)]
+        [LocationCompleter("Microsoft.RecoveryServices/vaults")]
+        [ValidateNotNullOrEmpty]
+        public string VaultLocation { get; set; }
```
Comment (maddieclayton) [inline src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/RSBackupVaultCmdletBase.cs]: Where is this used?

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Models/CmdletParamEnums.cs [hunk 7]
```diff
@@ -14,6 +14,13 @@
 
 namespace Microsoft.Azure.Commands.RecoveryServices.Backup.Cmdlets.Models
 {
+    public enum VaultParams
+    {
+        VaultName,
+        ResourceGroupName,
+        VaultLocation,
```
Comment (maddieclayton) [inline src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Models/CmdletParamEnums.cs]: I don't see this used anywhere - is it needed?

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/help/Backup-AzureRmRecoveryServicesBackupItem.md [hunk 8]
```diff
@@ -14,7 +14,7 @@ Starts a backup for a Backup item.
 ## SYNTAX
 
 ```
-Backup-AzureRmRecoveryServicesBackupItem -Item <ItemBase> [-ExpiryDateTimeUTC <DateTime>]
+Backup-AzureRmRecoveryServicesBackupItem -Item <ItemBase> [-ExpiryDateTimeUTC <DateTime>] [-Vault <ARSVault>]
```
Comment (maddieclayton) [inline src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/help/Backup-AzureRmRecoveryServicesBackupItem.md]: you will need to regenerate to get the new changes.

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/RSBackupVaultCmdletBase.cs (hunk 6 above)
Comment (dragonfly91) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup/RSBackupVaultCmdletBase.cs]: This is being used in the Restore cmdlet. You can find this change at https://github.com/Azure/azure-powershell/pull/6327/files#diff-b7f777ff31af1c21fbf6378a2f1e1647

File: src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Models/CmdletParamEnums.cs (hunk 7 above)
Comment (dragonfly91) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Models/CmdletParamEnums.cs]: This is being used in the Restore cmdlet. You can find this change at https://github.com/Azure/azure-powershell/pull/6327/files#diff-b7f777ff31af1c21fbf6378a2f1e1647

Comment (dragonfly91) [reply src/ResourceManager/RecoveryServices/Commands.RecoveryServices.Backup.Models/CmdletParamEnums.cs]: This is being used in the Restore cmdlet. https://github.com/Azure/azure-powershell/pull/6327/files#diff-b7f777ff31af1c21fbf6378a2f1e1647

**Comment by maddieclayton** [comment]
on demand here: https://azuresdkci.westus2.cloudapp.azure.com/view/PowerShell/job/powershell-demand/931/

**Comment by dragonfly91** [comment]
@maddieclayton Thank you for getting this in!
